### ISSUE TITLE
Benchmark and decide for the use of Proto3/FlatBuffer/CapnProto/DDS in OSI

### ISSUE THREAD

**Issue by HannesSchneiderAVL** [OpenSimulationInterface/open-simulation-interface#56]
We would like to use the OSI with csharp environment. But compiling classes for c# seems to be only supported for proto3. When compiling this version (changed the syntax variable to proto3) I get following errors:

osi_common.proto:14:14: Explicit 'optional' labels are disallowed in the Proto3
syntax. To define 'optional' fields in Proto3, simply remove the 'optional' labe
l, as fields are 'optional' by default.

for each definition.

then I removed the invalid optional statements.

and then following error occurs:

osi_common.proto: Explicit default values are not allowed in proto3.

by removing also the explicit values I was able to generete the class.

**Comment by CarloVanDriestenBMW** [comment]
Hello and welcome on board!
Please try to use a proto3 compiler but leave the syntax variable as "proto2". It just indicates the syntax of the files and does not restrict the usage of newer compiler (as far as I know :) )
I am curious if the C# compiler also works with proto2 syntax files, because the C++ compiler performs well (see e.g. the TravisCI log with proto3 compiler [link](https://travis-ci.org/OpenSimulationInterface/open-simulation-interface))

**Comment by pmai** [comment]
The original protobuf-csharp-port only supports proto2 and is still available, though in maintenance-only mode, see https://github.com/jskeet/protobuf-csharp-port

That code was merged into the mainline distribution and upgraded to proto3, however proto2 support was dropped as far as I can tell in the process:

http://stackoverflow.com/questions/36742747/proto2-vs-proto3-in-c-sharp

Discussions about moving to either Proto3, FlatBuffers or something like Cap'n'Proto would be a medium-term thing I think.

**Comment by HannesSchneiderAVL** [comment]
Thank you for your response. I have already upgraded to proto3. Unfortunately this is not possible without changing syntax variable and some other smaller modifications. Can I contact one of you for further details on the OSI interface? I could not find a contact email of Mr. Van Driesten. TY

**Comment by CarloVanDriestenBMW** [comment]
carlo.van-driesten@bmw.de
I will make it visible in my profile :)

**Comment by CarloVanDriestenBMW** [comment]
I will keep this issue open, adding the question of using flatbuffer/proto3 or captainProto

**Comment by CarloVanDriestenBMW** [comment]
@pmai 
Would it be a big difference to just use Proto3 for the next version?

**Comment by pmai** [comment]
It is probably a bit of effort, and I don't see that many benefits; I'd rather think about Flatbuffer support, which would seem to be a similar effort, with much bigger benefits...

**Comment by HannesSchneiderAVL** [comment]
A decission on this topic would be appreciated, because we are already building some prototypes/demonstrators based on OSI and a technology change would mean a high effort on our side.

**Comment by CarloVanDriestenBMW** [comment]
@HannesSchneiderAVL definitly! @pmai is it ok to use the newest ProtoCompiler for the proto2 syntax? I have the trouble that some other lib uses the proto 3.4 compiler?

**Comment by pmai** [comment]
@CarloVanDriestenBMW that should always work, you can even use proto2 message types inside proto3 message types.

**Comment by CarloVanDriestenBMW** [comment]
Proto3 has its own implementation for timestamps

**Comment by mbencik999** [comment]
I did a little research on the performance of the Protobuf serialization system and compared this with other systems.

https://codeburst.io/json-vs-protocol-buffers-vs-flatbuffers-a4247f8bda6f
https://steemit.com/programming/@boucaron/serialization-protocol-compilers
https://discuss.dgraph.io/t/comparison-of-protocol-buffers-v-s-flatbuffers/1085

Now I have to say I did not test and verify those results myself, thats why I quote 3 different sources. But those sources strongly suggest that the flatbuffer is very fast.

**Comment by HannesSchneiderAVL** [comment]
Are there any benchmark results regarding Protobuf available? or has anybody made experience with huge OSI packages (containing low level data)? Can someone estimate the effort in generating a flatbuf OSI version to do such a comparison?

**Comment by CarloVanDriestenBMW** [comment]
http://portals.omg.org/dds/why-choose-dds/

**Comment by jdsika** [comment]
https://www.obj-sys.com/products/asn1c/index.php

**Comment by ManuelKugelmann** [comment]
Newer protobuf c# versions have support for proto2. protobuf-net also supports proto2.

BTW: I did a simple Unity3D Plugin for OSI. I went with the google protobuf library because I was hoping that it would be very similar to the C++ variant, but that is not the case. I also had to generate some extra code and do some extensions to make it run GC free.  There is an open issue concerning the template factory in google protobuf for c# that could be fixed to simplify some of the GC avoidance code. An update to the newest protobuf might also be helpful.
protobuf-net might be a sensible alternative - it's more .NET affine.

Contact [usaneers ](https://www.usaneers.de/)if you are interested in that plugin. 
Contact [me ](https://www.xing.com/profile/Manuel_Kugelmann/cv)if you want your own implementation.
